rust-app/zephyr/src/uart.rs

```diff
@@ -0,0 +1,55 @@
+use crate::device::Device;
+
+pub trait UartSyscalls {
+    #[cfg(feature = "have_std")]
+    fn uart_poll_out(device: &Device, out_char: char);
+
+    #[cfg(feature = "have_std")]
+    fn uart_poll_in(device: &Device, in_char: &mut char) -> i32;
+
+    #[cfg(feature = "have_std")]
+    fn uart_err_check(device: &Device) -> i32;
+}
+
+macro_rules! trait_impl {
+    ($context:ident, $context_struct:path) => {
+        impl UartSyscalls for $context_struct {
+            #[cfg(feature = "have_std")]
+            #[inline(always)]
+            fn uart_poll_out(device: &Device, out_char: char) {
+                unsafe {
+                    zephyr_sys::syscalls::$context::uart_poll_out(
+                        core::mem::transmute(device),
+                        out_char as u8,
+                    )
+                };
+            }
+
+            #[cfg(feature = "have_std")]
+            #[inline(always)]
+            fn uart_poll_in(device: &Device, in_char: &mut char) -> i32 {
+                let mut munge: u8 = 0;
+                let rc: i32 = unsafe {
+                    zephyr_sys::syscalls::$context::uart_poll_in(
+                        core::mem::transmute(device),
+                        &mut munge,
+                    )
+                };
+                *in_char = munge as char;
+                rc
+            }
+
+            #[cfg(feature = "have_std")]
+            #[inline(always)]
+            fn uart_err_check(device: &Device) -> i32 {
+                (unsafe {
+                    zephyr_sys::syscalls::$context::uart_err_check(core::mem::transmute(device))
```

Comment:
Transmute works, but it's super dangerous so best to use normal casts where possible. Here you can do a cast from reference to const pointer, and then because it's unsafe, you can cast the cost pointer to a mut pointer. And _ can be used to infer the type.

`device as *const _ as *mut _`